### PR TITLE
HPCC-19447 Scope iterator failing if base scope given

### PR DESCRIPTION
**Pull request by jakesmith**
An error 'Logical filename cannot end with scope "."' was being
hit if a base scope was provided, picked up by validation added
in HPCC-18645.

The scope iterator was incorrectly adding a trailing ".".
It was also quite inefficient because even if a deep base scope
was provided it was always connecting from the root of the files.
Changed to connect directly.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Tested with daliadmin dfsscopes

### CODE REVIEW

**Comment by jakesmith** [comment]
@mckellyln - please review

File: dali/base/dadfs.cpp [hunk 1]
```diff
@@ -1904,22 +1904,23 @@ class CDFScopeIterator: implements IDFScopeIterator, public CInterface
...
-        StringBuffer baseq;
-        StringBuffer tmp;
-        if (base&&*base) {
+        StringBuffer lockPath;
+        if (!isEmptyString(base))
+        {
             CDfsLogicalFileName dlfn;
-            dlfn.set(base,".");
-            dlfn.makeScopeQuery(baseq,false);
+            dlfn.set(base, "dummyfilename"); // makeScopeQuery expects a lfn to a file, 'dummyfilename' will not be used
```
Comment (mckellyln) [inline dali/base/dadfs.cpp]: Don't understand the not be used comment

**Comment by mckellyln** [comment]
@jakesmith just one question about words in comment.

File: dali/base/dadfs.cpp (hunk 1 above)
Comment (jakesmith) [reply dali/base/dadfs.cpp]: dlfn.set wants a base scope path (e.g. scope1::scope2) and a trailing logical filename.
But in this context 'dlfn' is only used to get the xpath for the scope, the tail logicalname is not used, hence dummy.

**Comment by jakesmith** [comment]
@mckellyln - replied to comment.

**Review by mckellyln — APPROVED** [review]
@richardkchapman approved for merge

**Comment by ghalliday** [comment]
Looks good

**Pull request merged**